src/lustre/lustreTypeChecker.ml

```diff
@@ -1506,23 +1864,47 @@ let type_check_program: LA.t -> unit tc_result = fun prg ->
-      Log.log L_trace ("===============================================\n"
-                       ^^ "Phase 2: Type checking declaration Groups Done\n"
-                       ^^"===============================================\n")  
-      
-      ; report_tc_result tc_res
+      Log.log L_trace "Sorted consts and type decls:\n%a" LA.pp_print_program sorted_tys_consts
+
+      (* build the base context from the type and const decls *)
+      ; build_type_and_const_context empty_context sorted_tys_consts >>= fun global_ctx ->
+        Log.log L_trace ("===============================================\n"
+                         ^^ "Constant and type context \n"
+                         ^^ "TC Context\n%a\n"
+                         ^^"===============================================\n")
+          pp_print_tc_context global_ctx
+
+        (* Inline constants in constant declarations *)
+        ; Log.log L_trace ("===============================================\nStarting inlining constants") 
+        ; let (global_ctx', inlined_cs) = inline_constants global_ctx sorted_tys_consts in 
+          Log.log L_trace ( "Inlining constants done: \nConstant decls: %a"
+                            ^^"\n===============================================\n")
+            LA.pp_print_program inlined_cs
+
+          (* sort nodes and contract decls to make sure there are no recursive calls *)
+          ; AD.sort_declarations node_and_contract_decls >>= fun sorted_node_contract_decls -> 
+            
+            (* gather the typing context from nodes/functions and contracts *)
+            tc_context_of global_ctx' sorted_node_contract_decls >>= fun tc_ctx ->
+            Log.log L_trace ("===============================================\n"
+                             ^^ "Phase 1: Completed Building TC Global Context\n"
+                             ^^ "TC Context\n%a\n"
+                             ^^"===============================================\n")
+              pp_print_tc_context tc_ctx
+            ; (* type check the nodes and contract decls using this base typing context  *)
+              let tc_res = (type_check_decl_grps tc_ctx [prg]) in
+              Log.log L_trace ("===============================================\n"
+                               ^^ "Phase 2: Type checking declaration Groups Done\n"
+                               ^^"===============================================\n")  
+              (* inline constants in node body *)
+              ; let (_, sorted_node_contract_decls') = inline_constants global_ctx' sorted_node_contract_decls in
+                Log.log L_trace ("%a\n%a\n=============================================\n"
+                                 ^^ "Phase 3: Inline node constants done           \n"
+                                 ^^"===============================================\n")
+                  LA.pp_print_program inlined_cs
+                  LA.pp_print_program sorted_node_contract_decls'
```

Comment:
Why do you inline constants in node body here? You should move the code outside the typechecker, or include it in a later pull request.

Reply:
I think splitting it would be a better design.